### PR TITLE
Anon: Scaffold anon service

### PR DESCRIPTION
**Pull request by Jguer**
**What is this feature?**

- Remove experimental counts
- Add new table for anon devices
- Add basic methods for counting and list anon devices
 
**Special notes for your reviewer:**

Review commit per commit

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by Jguer** [comment]
@DanCech adding the scaffolding for a more permanent anon service. Happy to take any suggestions you might have mentioned async

File: pkg/services/anonymous/anonimpl/anonstore/database.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+			})
+			if err != nil {
+				s.log.Error("Failed to lock and execute cleanup old anon devices", "error", err)
+			}
+
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
```
Comment (kalleep) [inline pkg/services/anonymous/anonimpl/anonstore/database.go:148]: We are running this job every 2 hours but locking the execution of it to 10 hours, won't that cause some error logs when next couple of jobs are scheduled?

File: pkg/services/anonymous/anonimpl/anonstore/database.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	_, err := s.sqlStore.GetSqlxSession().Exec(ctx, "DELETE FROM anon_device WHERE device_id = ?", deviceID)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// deleteOldDevices deletes all devices that have no been updated since the given time.
+func (s *AnonDBStore) deleteOldDevices(ctx context.Context, olderThan time.Time) error {
+	_, err := s.sqlStore.GetSqlxSession().Exec(ctx, "DELETE FROM anon_device WHERE updated_at <= ?", olderThan.UTC())
```
Comment (kalleep) [inline pkg/services/anonymous/anonimpl/anonstore/database.go:122]: We could potentially have a lot of unique devices registered. Should we create and index on update_at?

Comment (kalleep) [reply pkg/services/anonymous/anonimpl/anonstore/database.go:122]: List would also benefit from that index

File: pkg/services/anonymous/anonimpl/impl.go [hunk 3]
```diff
@@ -2,178 +2,114 @@ package anonimpl
...
+	if err := a.anonStore.CreateOrUpdateDevice(ctx, device); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (a *AnonDeviceService) tagDeviceUI(ctx context.Context, httpReq *http.Request, device Device) error {
-	deviceID := httpReq.Header.Get(deviceIDHeader)
+func (a *AnonDeviceService) untagDevice(ctx context.Context,
```
Comment (kalleep) [inline pkg/services/anonymous/anonimpl/impl.go:93]: We should not untag a device if there was an error during the login proccess. Then a failed login would untag the device

File: pkg/services/anonymous/anonimpl/impl.go [hunk 4]
```diff
@@ -2,178 +2,118 @@ package anonimpl
...
-		IP:        device.IP,
-		UserAgent: device.UserAgent,
-	}); err != nil {
-		return err
-	}
-
-	return nil
 }
 
+// FIXME: Unexport and remove interface
```
Comment (kalleep) [inline pkg/services/anonymous/anonimpl/impl.go:110]: If you wan't to un export this you could just pass this as a function to the client :)

**Review by kalleep — APPROVED** [review]

File: pkg/services/anonymous/anonimpl/impl.go (hunk 4 above)
Comment (Jguer) [reply pkg/services/anonymous/anonimpl/impl.go:110]: I kept it as a fixme as I think this PR is already a bit over dimensioned as is :), makes a good beginner 'get to know wire' task

**Review by DanCech — APPROVED** [review]
Nice work!

**Comment by danielkenlee** [comment]
@vtorosyan - I approved to get this backported to all versions back to v9.5.x. This is required to more accurately count the [Customers with anonymous auth enabled](https://ops.grafana-ops.net/d/DLdBlQN4k/authnz-usage-stats?orgId=1&viewPanel=21)

- Jo attempted to get this ported a couple of weeks ago, but looks like there's been some failures. 
- It would be useful to already start collecting data from the enterprise customers using the latest counting method in this PR.
- Would another member of the team be able to step in to backport manually? I saw @kalleep was one of the reviewers.
- If there are too many conflicts to resolve then we can wait till Jo returns, but it would be good to not miss out on a few weeks of data.

**Comment by kalleep** [comment]
@danielkenlee This one is a bit tricky to backport and a lot of functionality would have to be re-implemented.

**Comment by vtorosyan** [comment]
@kalleep we just talked with @alexanderzobnin about this and he will be taking this over.

@danielkenlee sorry that the backports did not get through. we will try to prioritize this week. Note that for the backports to be released we need to make an official request, so when the backports are ready team will get back to you for adding information about why we need to release the change (as typically changes which are not bug fixes are not being released with the previous versions).

**Pull request merged**